File: 1/2021.03.29/990.Satisfiability of Equality Equations/python/solution.py
class Solution:

    def equationsPossible(self, equations: List[str]) -> bool:
        uf = UnionFind(26)
        for s in equations:
            if s[1] == '=':
                uf.union(ord(s[0]) - ord('a'), ord(s[3]) - ord('a'))
        for s in equations:
            if s[1] == '!':
                if uf.connected(ord(s[0]) - ord('a'), ord(s[3]) - ord('a')):
                    return False
        return True


class UnionFind:

    def __init__(self, n) -> None:
        self.count = n
        self.parent = [i for i in range(n)]
    
    def find(self, x):
        while x != self.parent[x]:
            self.parent[x] = self.parent[self.parent[x]]
            x = self.parent[x]
        return x
    
    def union(self, p, q):
        root_p = self.find(p)
        root_q = self.find(q)
        if root_p == root_q:
            return
        self.parent[root_p] = root_q
        self.count -= 1
    
    def connected(self, p, q):
        root_p = self.find(p)
        root_q = self.find(q)
        return root_p == root_q

File: 1/2021.03.29/990.Satisfiability of Equality Equations/java/solution.java
class Solution {
    public boolean equationsPossible(String[] equations) {
        UnionFind uf = new UnionFind(26);
        List<String> noEqual = new ArrayList<>();
        for (String s : equations) {
            if (s.charAt(1) == '=') {
                uf.union(s.charAt(0) - 'a', s.charAt(3) - 'a');
            } else {
                noEqual.add(s);
            }
        }
        for (String s : noEqual) {
            if (uf.connected(s.charAt(0) - 'a', s.charAt(3) - 'a')) {
                return false;
            }
        }
        return true;
    }

    class UnionFind {
        private int count;
        private int[] parent;

        public UnionFind(int n) {
            this.count = n;
            parent = new int[n];
            for (int i = 0; i < n; i++) {
                parent[i] = i;
            }
        }

        public int find(int x) {
            while (parent[x] != x) {
                parent[x] = parent[parent[x]];
                x = parent[x];
            }
            return x;
        }

        public void union(int p, int q) {
            int rootP = find(p);
            int rootQ = find(q);
            if (rootP == rootQ) {
                return;
            }
            parent[rootP] = rootQ;
            count--;
        }

        public boolean connected(int p, int q) {
            int rootP = find(p);
            int rootQ = find(q);
            return rootP == rootQ;
        }
    }
}
